### ISSUE TITLE
mission download on the server side

### ISSUE THREAD

**Issue by heiheiheihhhhh** [mavlink/MAVSDK#2340]
Hi,If I am not directly connected to PX4 or AUTOPILOT, does the mission download on the server side have an interface?

**Comment by julianoes** [comment]
I don't understand what you mea. Can you give a bit more context about your setup and what you're trying to do?

**Comment by julianoes** [comment]
Aha, so on the server (so autopilot side) you need to use the [MissionRawServer](https://github.com/mavlink/MAVSDK/blob/main/src/mavsdk/plugins/mission_raw_server/include/plugins/mission_raw_server/mission_raw_server.h) API.

We call upload from ground station to autopilot, by the way.

Also, if you use server plugins, make sure to use MAVSDK v2. Server plugins were not well supported with MAVSDK v1.4.

**Comment by heiheiheihhhhh** [comment]
> Aha, so on the server (so autopilot side) you need to use the [MissionRawServer](https://github.com/mavlink/MAVSDK/blob/main/src/mavsdk/plugins/mission_raw_server/include/plugins/mission_raw_server/mission_raw_server.h) API.
> 
> We call upload from ground station to autopilot, by the way.
> 
> Also, if you use server plugins, make sure to use MAVSDK v2. Server plugins were not well supported with MAVSDK v1.4.

Thank you for your reply. The situation is that I can currently implement the flight paths upload from the ground station to autopilot. I want to implement the flight paths download from the autopilot to the ground station and have found the mission_download API in the mission plugin, but I am not sure how to provide the flight paths to the ground station.

**Comment by julianoes** [comment]
Plugins used on the autopilot side are called `_server`. So that's what you want to use on that side, generally.

**Comment by heiheiheihhhhh** [comment]
> Plugins used on the autopilot side are called . So that's what you want to use on that side, generally.`_server`

But there is no download_mission API in mission_raw_server plugin.

**Comment by julianoes** [comment]
No, the way it works is that the client (e.g. a ground station) does the upload or download. The server just responds and does the transfer.

**Comment by heiheiheihhhhh** [comment]
> No, the way it works is that the client (e.g. a ground station) does the upload or download. The server just responds and does the transfer.

I have tried using the mission_download API of the mission plugin on the client side, but it returned an empty MissionPlan.

**Comment by julianoes** [comment]
So there was no mission saved? Sorry, it's difficult to understand what is going on. You might have to share a code snippet or reproducible example.

**Comment by heiheiheihhhhh** [comment]
> So there was no mission saved? Sorry, it's difficult to understand what is going on. You might have to share a code snippet or reproducible example.

I apologize for the delayed response. I have reviewed the MAVLink official website regarding the download sequence diagram and found that in MAVSDK, there is only the send of MISSION_REQUEST_LIST, with no subscribe. In other words, MAVSDK indeed lacks the communication for download on the server side. Therefore, I have added a class called DownloadServerWorkItem to the mavlink_mission_transfer framework to solve this problem.
![download](https://github.com/mavlink/MAVSDK/assets/146074081/bd741194-7248-4c84-afba-a3e7e828c331)
![transfer](https://github.com/mavlink/MAVSDK/assets/146074081/67a18304-62f4-49d9-b169-49506e9d7b56)

**Comment by julianoes** [comment]
I think you're misunderstanding this. The server doesn't initiate or do a mission download. It only "serves". If you want to download a mission, you would use the mission_raw plugin.

**Comment by heiheiheihhhhh** [comment]
> I think you're misunderstanding this. The server doesn't initiate or do a mission download. It only "serves". If you want to download a mission, you would use the mission_raw plugin.

I have tried to directly call mission_upload and mission_download in the mission plugin, but I did not obtain the MissionPlan. Below are parts of my source code and the results of execution.
<img width="813" alt="run" src="https://github.com/mavlink/MAVSDK/assets/146074081/0f31af63-6fbd-4385-b668-23fb13e555cb">
<img width="830" alt="server" src="https://github.com/mavlink/MAVSDK/assets/146074081/e5ab0428-5e76-45a2-9db2-e367c18b581f">
<img width="875" alt="client" src="https://github.com/mavlink/MAVSDK/assets/146074081/12102798-8b25-4892-975c-982107590971">

**Comment by julianoes** [comment]
Ok, that would be a bug. Can you share code for me to reproduce and also mention PX4 or ArduPilot version that you're using this against?

**Comment by julianoes** [comment]
OP never cam back with an example, closing.
